src/client/opamClient.ml

```diff
@@ -202,9 +202,25 @@ let upgrade_t
         else OpamPackage.packages_of_names t.installed requested
       in
       let latest =
-        OpamPackage.Name.Set.fold (fun name acc ->
-            OpamPackage.Set.add (OpamPackage.max_version t.packages name) acc)
-          (OpamPackage.names_of_packages to_check)
+        OpamPackage.Set.fold (fun pkg acc ->
+            let name = OpamPackage.name pkg in
+            let pkgs = OpamPackage.packages_of_name t.packages name in
+            let latest =
+              OpamPackage.Set.fold (fun pkg latest ->
```

Comment:
this variable named `pkg` and the one defined 4 lines above are not the "same package" but have the same name, this can be misleading when reading the code.